app/controllers/incomeSources/add/ForeignPropertyReportingMethodController.scala

```diff
@@ -0,0 +1,257 @@
+                                                        (implicit val appConfig: FrontendAppConfig,
+                                                         override implicit val mcc: MessagesControllerComponents,
+                                                         val ec: ExecutionContext,
+                                                         val itvcErrorHandler: ItvcErrorHandler,
+                                                         val itvcErrorHandlerAgent: AgentItvcErrorHandler
+                                                        )
+  extends ClientConfirmedController with FeatureSwitching with I18nSupport {
+
+  def submit(id: String): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleSubmitRequest(
+        id = id,
+        isAgent = false,
+        itvcErrorHandler = itvcErrorHandler,
+        redirectCall = redirectCall(id),
+        postAction = postAction(id)
+      )
+  }
+
+  def submitAgent(id: String): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService).flatMap {
+          implicit mtdItUser =>
+            handleSubmitRequest(
+              id = id,
+              isAgent = true,
+              itvcErrorHandler = itvcErrorHandlerAgent,
+              redirectCall = redirectCallAgent(id),
+              postAction = postActionAgent(id)
+            )
+        }
+  }
+
+  def show(id: String): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(
+        id = id,
+        isAgent = false,
+        itvcErrorHandler = itvcErrorHandler,
+        redirectCall = redirectCall(id),
+        postAction = postAction(id),
+      )
+  }
+
+  def showAgent(id: String): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService).flatMap {
+          implicit mtdItUser =>
+            handleRequest(
+              id = id,
+              isAgent = true,
+              itvcErrorHandler = itvcErrorHandlerAgent,
+              redirectCall = redirectCallAgent(id),
+              postAction = postActionAgent(id),
+            )
+        }
+  }
+
+  private def handleRequest(id: String,
+                            isAgent: Boolean,
+                            itvcErrorHandler: ShowInternalServerError,
+                            postAction: Call,
+                            redirectCall: Call)
+                           (implicit user: MtdItUser[_]): Future[Result] = {
+
+    if (isEnabled(IncomeSources)) {
+      itsaStatusService.hasMandatedOrVoluntaryStatusCurrentYear.flatMap {
```

Comment:
* pls use for-comprehension in this case, as its more readable, there are few nesting pattern-matching, so its hard to understand if you missed any case or not; 
* Here is relevant compilation warning you need to address:

` /Users/hmrc/devcore/itvc/income-tax-view-change-frontend/app/controllers/incomeSources/add/ForeignPropertyReportingMethodController.scala:122:73: match may not be exhaustive.
[warn] It would fail on the following input: false
[warn]       itsaStatusService.hasMandatedOrVoluntaryStatusCurrentYear.flatMap {
[warn]                                                                         ^`